src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs

```diff
@@ -19,6 +19,31 @@ namespace System
     /// </summary>
     public static partial class MemoryExtensions
     {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static bool SliceArgumentsAreValid(int sourceLength, int start, int length)
+        {
+#if TARGET_64BIT
+            // Since start and length are both 32-bit, their sum can be computed across a 64-bit domain
+            // without loss of fidelity. The cast to uint before the cast to ulong ensures that the
+            // extension from 32- to 64-bit is zero-extending rather than sign-extending. The end result
+            // of this is that if either input is negative or if the input sum overflows past Int32.MaxValue,
+            // that information is captured correctly in the comparison against the source length.
+            // We don't use this same mechanism in a 32-bit process due to the overhead of 64-bit arithmetic.
+            return (ulong)(uint)start + (ulong)(uint)length <= (ulong)(uint)sourceLength;
+#else
+            return (uint)start <= (uint)sourceLength && (uint)length <= (uint)(sourceLength - start);
+#endif
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void ValidateSliceArguments(int sourceLength, int start, int length)
+        {
+            if (!SliceArgumentsAreValid(sourceLength, start, length))
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException();
```

Comment:
We've actually gotten several complaints over us not reporting the different reasons for `ArgumentOutOfRangeException` accurately.

Centralizing things is good and makes it easier to ensure consistency. However, I think we may want to have the cheap early check and then have a `ThrowHelper.ThrowArgumentOutOfRangeExceptionForInvalidSlice(sourceLength, start, length)` which goes and determines which input was actually bad so we can provide accurate failure info as the way to be consistent.

Reply:
Past discussions have leaned towards removing such checks instead: https://github.com/dotnet/runtime/issues/115239#issuecomment-2847426917, https://github.com/dotnet/runtime/issues/53622#issuecomment-853284509

Do you think it'd be worth extra cost on all call sites given we're talking about `Span`?

Reply:
> However, I think we may want to have the cheap early check and then have a `ThrowHelper.ThrowArgumentOutOfRangeExceptionForInvalidSlice(sourceLength, start, length)` which goes and determines which input was actually bad so we can provide accurate failure info as the way to be consistent.

An argument against this is we would have to pass the parameter names in order to report correctly, e.g. `start` could be `start`, `startIndex`, `offset`, `desiredStartIndex`, etc.